glide-core/src/client/value_conversion.rs

```diff
@@ -673,6 +674,289 @@ pub(crate) fn convert_to_expected_type(
+        18) 1) "1719710679916-0"
+            2) 1) "foo"
+               2) "bar"
+               3) "foo"
+               4) "bar2"
+               5) "some"
+               6) "value"
+        19) "last-entry"
+        20) 1) "1719710688676-0"
+            2) 1) "foo"
+               2) "bar2"
+
+        RESP3 on an empty stream:
+
+        1# "length" => (integer) 0
+        2# "radix-tree-keys" => (integer) 0
+        3# "radix-tree-nodes" => (integer) 1
+        4# "last-generated-id" => "0-1"
+        5# "max-deleted-entry-id" => "0-1"
+        6# "entries-added" => (integer) 1
+        7# "recorded-first-entry-id" => "0-0"
+        8# "groups" => (integer) 0
+        9# "first-entry" => (nil)
+        10# "last-entry" => (nil)
+
+        So we convert:
+        - Arrays to maps, accroding to RESP2 -> RESP3 conversion done by the server:
+          - Top level array - unflat an `obj[]` (`[k, v, k, v, ...]`) to a `Map<str, obj>`
+          - "groups" value - `obj[][]` -> `Map<str, obj>[]` - unflat every nested array the same manner as above
+          - "consumers" value - `obj[][]` -> `Map<str, obj>[]` - same as above
+        - Additionally we convert some map's values:
+          - "entries", "first-entry" and "last-entry" value - to a `Map<str, str[][]>` (similar to `XREAD`)
+              no nested maps due to duplicating keys - see example
+        Using `XInfoStreamFullReturnType` recursively for maps' value type.
+        */
+        ExpectedReturnType::XInfoStreamFullReturnType => match value {
+            Value::Map(_) => Ok(value),  // Response is already in RESP3 format - no conversion needed
+            Value::Array(mut array) => {
+                // Response is in RESP2 format. We need to:
+                // - convert any consumer in the consumer array to a map, if there are any consumers
+                // - convert any group in the group array to a map, if there are any groups
+                // - convert the root of the response into a map
+                let groups_key = Value::SimpleString("groups".into());
+                let opt_groups_index = array
+                    .iter()
+                    .position(
+                        |key| {
+                            let res = convert_to_expected_type(key.clone(), Some(ExpectedReturnType::SimpleString));
+                            match res {
+                                Ok(converted_key) => {
+                                    converted_key == groups_key
+                                },
+                                Err(_) => {
+                                    false
+                                }
+                            }
+                        }
+                    );
+
+                let Some(groups_index) = opt_groups_index else {
+                    return Err((ErrorKind::TypeError, "Groups key not found").into());
+                };
+
+                if array.get(groups_index + 1).is_none() {
```

Comment:
Not sure if we need it here, but you might want to add an overflow check here, in case groups_index is usize::MAX

Reply:
Definitely a good callout, but I don't think it is needed. usize::MAX is `18_446_744_073_709_551_615`, but groups_index + 1 is  always 18 with the current version of redis. For this to overflow, valkey would have to add enough fields in this array to boost that number over usize::MAX, which seems very unlikely.